application/src/main/java/org/togetherjava/tjbot/commands/moderation/WarnCommand.java

```diff
@@ -0,0 +1,193 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.togetherjava.tjbot.commands.SlashCommandAdapter;
+import org.togetherjava.tjbot.commands.SlashCommandVisibility;
+import org.togetherjava.tjbot.db.Database;
+import org.togetherjava.tjbot.db.DatabaseException;
+import org.togetherjava.tjbot.db.generated.tables.Warns;
+import org.togetherjava.tjbot.db.generated.tables.records.WarnsRecord;
+
+import java.awt.*;
+import java.util.Objects;
+import java.util.Optional;
+
+public class WarnCommand extends SlashCommandAdapter {
+    private static final Logger logger = LoggerFactory.getLogger(WarnCommand.class);
+    private static final String WARN_USER = "warn_user";
+    private static final String WARN_USER_OPTION = "user";
+    private static final String WARN_REASON_OPTION = "reason";
+    private static final String RETRIEVE_WARNS_OPTION = "retrieve_warns";
+    private static final String RETRIEVE_USER_OPTION = "user";
+    private final Database database;
+
+    /**
+     * Creates a new adapter with the given data.
+     *
+     * @param database the database to store the key-value pairs in
+     */
+    public WarnCommand(@NotNull Database database) {
+        super("warn", "warns the user", SlashCommandVisibility.GUILD);
+        this.database = database;
+
+        getData().addSubcommands(new SubcommandData(WARN_USER, "Used to warn the user")
+            .addOption(OptionType.USER, WARN_USER_OPTION, "The user to warn", true)
+            .addOption(OptionType.STRING, WARN_REASON_OPTION, "The reason for the warning", true),
+                new SubcommandData(RETRIEVE_WARNS_OPTION, "Uses to retrieve a warn for a member")
+                    .addOption(OptionType.USER, RETRIEVE_USER_OPTION,
+                            "The warns for the user you want", true));
+    }
+
+    @Override
+    public void onSlashCommand(@NotNull SlashCommandEvent event) {
+        switch (Objects.requireNonNull(event.getSubcommandName())) {
+            case WARN_USER -> handleWarnCommand(event);
+            case RETRIEVE_WARNS_OPTION -> handleRetrieveWarnCommand(event);
+            default -> throw new AssertionError();
+        }
+    }
+
+    /**
+     * Handles {@code /warn retrieve_warns user} commands. Retrieves the timesWarned value and then
+     * sends that value to the user as an embed.
+     * <p>
+     * This command can only be used by users with the {@code BAN_MEMBERS} permission.
+     *
+     * @param event the event of the command
+     */
+    private void handleRetrieveWarnCommand(@NotNull CommandInteraction event) {
+        OptionMapping userOption =
+                Objects.requireNonNull(event.getOption(RETRIEVE_USER_OPTION), "The target is null");
+        Member author = userOption.getAsMember();
+
+        // To prevent people from get users warn, only users with
+        // elevated permissions are allowed to use this command
+        if (handleHasPermissions(author, event, event.getGuild()).equals(false)) {
```

Comment:
`.equals(false)` what is this? if you want to negate the condition, do `if (!...)`